TDL11-Terminate_Instances_with_snapshots/terminate_aged_instances_with_snapshots.py

```diff
@@ -0,0 +1,224 @@
+
+# Default time_zone, current_time
+region_name = 'ap-southeast-2'
+time_zone = 'Australia/Melbourne'
+tz = pytz.timezone(time_zone)
+current_time = datetime.datetime.now(tz).strftime('%Y%m%d%H%M') # if not provided, set current_time to default_current_time
+
+tag_key='AgedTime'
+
+# check current_time is provided from command line, and format is correct
+def validate_datetime_format(date_text,date_formate,date_input_format):
+    try:
+        datetime.datetime.strptime(date_text,date_formate)
+    except ValueError:
+        raise ValueError("Incorrect data format, should be "+date_input_format)
+
+# Check EC2 instances in your accouunt which has tag AgedTime and AgedTime <= current_time
+def get_instances_by_tag(tag_key,current_time):
+    response = ec2.describe_instances(
+        Filters=[
+            {
+                'Name': 'tag:'+tag_key,
+                'Values': [
+                    '*'
+                ]
+            } 
+        ]
+    )
+    instanceIds = []
+    instances = []
+    for reservation in response["Reservations"]:
+        for instance in reservation["Instances"]:
+            InstanceId = (instance['InstanceId'])
+            ec2instance = resource.Instance(InstanceId)
+            for tags in ec2instance.tags:
+                if tags["Key"] == tag_key:
+                    agedtimeValue = tags["Value"]
+                    print(f'agedtime:{agedtimeValue}')
+                    if agedtimeValue <= current_time: 
+                        print('agedtimeValue <= current_time')
+                        instanceIds.append(instance["InstanceId"])
+                        instances.append(instance)
+    #print(json.dumps(instanceIds, indent=5))
+    #print(instanceIds, sep = "\n") 
+    pprint.pprint(instanceIds)
+    return instances
+
+def shutdown_instance(instance_id):
+    print(f"Instance is running, shutting down prior to creating snapshot(s) of attached volume(s)")
+    # Do a dryrun first to verify permissions
+    try:
+        ec2.stop_instances(InstanceIds=[instance_id], DryRun=True)
+    except ClientError as e:
+        if 'DryRunOperation' not in str(e):
+            raise e
+    # Dry run succeeded, call stop_instances without dryrun
+    try:
+        response = ec2.stop_instances(InstanceIds=[instance_id], DryRun=False)
+        logger.info("Stopping instance %s.", instance_id)
+        print(response)
+    except ClientError as e:
+        logger.exception("Couldn't stop instance %s.", instance_id)
+        raise e
+    print(f"Waiting for Instance {instance_id} to be shutdown")
+    shutdown_instance_wait(instance_id)
+
+def shutdown_instance_wait(instance_id):
+    shutdown_instance_waiter = ec2.get_waiter('instance_stopped')
+    try:
+        shutdown_instance_waiter.wait(InstanceIds=[instance_id])
+        print(f"Instance {instance_id} has shutdown successfully")
+    except botocore.exceptions.WaiterError as e:
+        print(e)
+
+def snapshots_wait(snapshot_ids):
```

Comment:
For lambda version,  we should try to avoid wait async operations such as stop/start instances, snapshot creation for the below reasons:
- We are paying the running time of the lambda function 
- In your case, the lambda function will run every 5 minutes, so we can just stop the instance, start snapshot creation, terminate the instance(terminate instance might fail if a snapshot is taking, but will succeed next time.

In this case, we can  take snapshots, terminate instances with minimum cost and multiple run.

Reply:
OK, thanks.